src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/RazorCodeDocument.cs

```diff
@@ -60,140 +74,198 @@ public static RazorCodeDocument Create(
     {
         ArgHelper.ThrowIfNull(source);
 
-        return new RazorCodeDocument(source, imports, parserOptions, codeGenerationOptions);
+        return new RazorCodeDocument(
+            source,
+            imports,
+            parserOptions ?? RazorParserOptions.Default,
+            codeGenerationOptions ?? RazorCodeGenerationOptions.Default,
+            tagHelpers: null,
+            referencedTagHelpers: null,
+            preTagHelperSyntaxTree: null,
+            syntaxTree: null,
+            importSyntaxTrees: default,
+            tagHelperContext: null,
+            documentNode: null,
+            csharpDocument: null);
     }
 
     internal bool TryGetTagHelpers([NotNullWhen(true)] out TagHelperCollection? result)
-        => _properties.TagHelpers.TryGetValue(out result);
+    {
+        result = _tagHelpers;
+        return result is not null;
+    }
 
     internal TagHelperCollection? GetTagHelpers()
-        => _properties.TagHelpers.Value;
+        => _tagHelpers;
 
     internal TagHelperCollection GetRequiredTagHelpers()
-        => _properties.TagHelpers.RequiredValue;
+        => _tagHelpers.AssumeNotNull();
 
-    internal void SetTagHelpers(TagHelperCollection? value)
-        => _properties.TagHelpers.SetValue(value);
+    internal RazorCodeDocument WithTagHelpers(TagHelperCollection? value)
+    {
+        if (ReferenceEquals(value, _tagHelpers))
```

Comment:
`TagHelperCollection` has efficient value equality. So, it's probably better to do `Equals(value, _tagHelpers)`.